core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceCreator.scala

```diff
@@ -0,0 +1,688 @@
+      _ <- syncWorkspacePoliciesWithGoogle(workspace, policyNameToEmailMap, authDomains, span)
+      _ <- setupGoogleProjectForWorkspace(workspace, policyNameToEmailMap, billingProjectOwnerPolicyEmail, authDomains, span)
+      _ <- maybeUpdateGoogleProjectsInPerimeter(billingProject)
+    } yield workspace
+  }
+
+  /**
+    * Creates a new Workspace doing everything the same as createWorkspaceInternal_sequential, but this version tries
+    * to run certain steps/futures in parallel instead of running them all sequentially.  It's not super pretty.  It
+    * breaks things down into 3 different steps, each of which is its own `for` comprehension.  Why 3?  It was chosen
+    * because there was a set of steps identified that could be run first and all in parallel.  After these finish,
+    * there is a set of steps that must be executed sequentially.  After those complete, there is another set of steps
+    * that can all be kicked off in parallel.  If you need to add a new step, depending on its placement and whether it
+    * needs to be run in parallel or sequentially with existing steps, you may be able to just add it to an existing
+    * `for` comp, or you may need to add a whole new `for` comp.
+    * @param workspaceRequest
+    * @param span
+    * @return
+    */
+  private def createWorkspaceInternal_parallel(workspaceRequest: WorkspaceRequest, span: Span = null): Future[Workspace] = {
+    val authDomains = workspaceRequest.authorizationDomain.getOrElse(Set.empty)
+    val noWorkspaceOwner = workspaceRequest.noWorkspaceOwner.getOrElse(false)
+
+    for {
+      (billingProjectOwnerPolicyEmail, billingProject) <- firstSteps(workspaceRequest, span)
+      (workspace, workspaceResourceResponse) <- secondSteps(workspaceRequest, billingProject, billingProjectOwnerPolicyEmail, noWorkspaceOwner, authDomains, span)
+      _ <- thirdSteps(workspace, billingProject, workspaceResourceResponse, billingProjectOwnerPolicyEmail, authDomains, span)
+    } yield workspace
+  }
+
+  private def firstSteps(workspaceRequest: WorkspaceRequest, span: Span = null): Future[(WorkbenchEmail, RawlsBillingProject)] = {
+    val checkUserCanCreateWorksapce = checkCreateWorkspacePermissions(workspaceRequest, span)
+    val checkWorkspaceNameIsUnique = checkIfWorkspaceWithNameAlreadyExists(workspaceRequest.toWorkspaceName, span)
+    val billingProjectOwnerPolicyEmailFuture = getBillingProjectOwnerPolicyEmail(workspaceRequest, span)
+    val getBillingProjectFuture = getBillingProjectForNewWorkspace(workspaceRequest.namespace, span)
+
+    for { // These things happen in parallel, but they all need to finish before moving on
+      _ <- checkUserCanCreateWorksapce
+      _ <- checkWorkspaceNameIsUnique
+      billingProjectOwnerPolicyEmail <- billingProjectOwnerPolicyEmailFuture
+      billingProject <- getBillingProjectFuture
+    } yield (billingProjectOwnerPolicyEmail, billingProject)
+  }
```

Comment:
an interesting side effect of running these in parallel is the possibility of information leakage. Consider a user who has zero permissions to create a workspace in billing project X. Now consider that user is trying to create a workspace named "foo". The user can't see that a workspace X/foo exists, because they don't have permissions.

If `checkCreateWorkspacePermissions` is slow because it's querying Sam, but `checkIfWorkspaceWithNameAlreadyExists` is fast because it's only going to the local database ... the user could end up with an error message of "Workspace ${workspaceName} already exists", when otherwise they would not have been able to know if that workspace existed or not.

Reply:
yeah, totally, and this actually did trigger a test failure.  I put things back to be mostly sequential for now.  Hopefully in the future we can find ways to parallelize this more.